linalg/operator.hpp

```diff
@@ -442,6 +442,133 @@ class TimeDependentOperator : public Operator
    virtual ~TimeDependentOperator() { }
 };
 
+
+
+/**
+  TimeDependentAdjointOperator is a TimeDependentOperator with Adjoint rate
+  equations to be used with CVODESSolver
+*/
+class TimeDependentAdjointOperator : public TimeDependentOperator
+{
+public:
+
+   /**
+      \brief The TimedependentAdjointOperator extends the TimeDependentOperator
+      class to use features in SUNDIALS CVODESSolver for computing quadratures
+      and solving adjoint problems.
+
+      To solve adjoint problems one needs to implement the AdjointRateMult method
+       to tell CVODES what the adjoint rate equation is.
+
+      QuadratureIntegration (optional) can be used to compute values over the
+      forward problem
+
+      QuadratureSensitivityMult (optional) can be used to find the sensitivity of
+      the quadrature using the adjoint solution in part.
+
+      SUNImplicitSetupB (optional) can be used to setup custom solvers for the
+      newton solve for the adjoint problem.
+
+      SUNImplicitSolveB (optional) actually uses the solvers from
+      SUNImplicitSetupB to solve the adjoint problem.
+
+      See SUNDIALS user manuals for specifics.
+    **/
+
+   TimeDependentAdjointOperator(int dim, int adjdim, double t = 0.,
+                                Type type = EXPLICIT) :
+      TimeDependentOperator(dim, t, type),
+      adjoint_height(adjdim)
+   {}
```

Comment:
I hope the goggly eyes don't expose my ignorance here :laughing:

Reply:
I don't actually know what they are for. I just wanted to tell you "I'm looking at it".

Reply:
So there is a bit of a backstory to this. The reason why I wrote it this way is that because the original SUNDIALS advection diffusion example actually uses the adjoint solution to compute the quadrature in the miniapp using (NEQ+NP) values vs (NEQ) primal variables. 

I took the liberty to rewrite the example just to showcase the quadrature calls better. So CVODES can indeed calculate adjoints that are different sizes compared to the primal, but the implementation becomes more convoluted.

The philosophical question is whether we should prohibit it's use via the interface, or allow people to choose themselves. I don't have a strong opinion on this, but of course I'd prefer to have the option to change the size without having to modify the code.

@gardner48 : Can you comment on this?

Reply:
Using the quadrature capability is the preferred approach so I think its use should be supported by the interface and let user choose what they would prefer.

Reply:
We've actually been meaning to update that example to use the quadrature calls but have not gotten around to it yet.

Reply:
I see the point. Can you put a quick note in the documentation?